community/kernel/src/main/java/org/neo4j/kernel/impl/util/MonotonicCounter.java

```diff
@@ -32,9 +32,17 @@ static MonotonicCounter newAtomicMonotonicCounter()
             private final AtomicLong value = new AtomicLong( 0L );
 
             @Override
-            public long incrementAndGet()
+            public long increment()
             {
-                return value.incrementAndGet();
+                int initialValue;
+                int incrementedValue;
+                do
+                {
+                    initialValue = value.intValue();
+                    incrementedValue = initialValue < 0 ? 0 : initialValue + 1;
+                }
+                while ( !value.compareAndSet( initialValue, incrementedValue ) );
+                return incrementedValue;
```

Comment:
This will produce a sequence of 1, 2, 3 … Integer.MAX_VALUE, Integer.MIN_VALUE, 0, 1, 2, 3 …

Looks like a bug. Why not just keep the old implementation?

Reply:
bug fixed, test introduced. I want to make sure that tx ids or query ids are always correct and guarantee that they will never be negative, even in this unlikely, but still possible case.

Reply:
no, you are obviously right that's too unrealistic as well will restore original counter thingy